pkg/resource/deploy/source_eval.go

```diff
@@ -2197,7 +2227,28 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 			Options: options,
 		})
 		if err != nil {
-			return nil, err
+			if st, ok := status.FromError(err); ok {
+				// We only turn errors into diagnostics for a few specific status codes that we know provider
+				// constructs return, and that aren't handled by the SDK.  The SDKs treat some status codes
+				// specially, and we want to keep that special handling, so we need to return the error verbatim
+				// to the SDK.
+				//nolint:exhaustive // Remaining cases are covered by default case.
+				switch st.Code() {
+				case codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.FailedPrecondition, codes.Aborted,
+					codes.OutOfRange, codes.DataLoss:
+					message := statusToMessage(st)
+					rm.diagnostics.Errorf(diag.GetResourceInvalidError(constructResult.URN), t, name, message)
```

Comment:
Again, will this do nothing if the message that comes back is empty? Or do we emit a blank line/empty diagnostic? Feels like it'd be good to avoid that if we do.

Reply:
Bah my bad, I totally missed this comment earlier.  It does look like it will emit an empty diagnostic if we don't pass it a message.  That said though, there should always be a message in the status (this is going to be a non-optional in the error class.  Nothing that prevents people from passing empty strings, but they shouldn't).  Even if there is no message, the type and name is still valid, so it's going to be an error like:

```
<t> resource '<name>' has a problem:
```

A little ugly, but still better than failing and not printing an error at all imo.